static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts

```diff
@@ -41,101 +48,67 @@ const shouldRenderAdTestLabel = (): boolean =>
+	adTestCookieRemovalLink.style.cssText =
+		'position: relative;padding: 0 0.5rem;text-align: left;box-sizing: border-box;';
 
+	if (shouldRender && val) {
 		const url = new URL(window.location.href);
 		url.searchParams.set('adtest', 'clear');
-
 		const clearLink = document.createElement('a');
+		clearLink.className = 'ad-slot__adtest-cookie-clear-link';
 		clearLink.href = url.href;
 		clearLink.innerHTML = 'clear';
-		adTestLabel.appendChild(clearLink);
+		adTestCookieRemovalLink.appendChild(clearLink);
 	}
 
-	return adTestLabel;
-};
-
-const createAdLabel = (): HTMLElement => {
-	const adLabel = document.createElement('div');
-	adLabel.className = 'ad-slot__label';
-	adLabel.innerHTML = 'Advertisement';
-	adLabel.appendChild(createAdTestLabel());
-	adLabel.appendChild(createAdCloseDiv());
-	return adLabel;
+	return adTestCookieRemovalLink;
 };
 
 /**
- *  **Dynamic labels:**
- *  Advert labels are historically inserted dynamically as a child of the advert slot node.
- *  This causes a cumulative layout shift (CLS) as content below is pushed down. This is
- *  particularly noticeable when ads are refreshed as the advert slot contents are deleted.
- *
- *  **Toggled labels:**
- *  To prevent CLS the label inserted on the server with its visibility initially hidden.
- *  Its visibility and width is toggled once the ad and its width is known.
- *  Currently only for dfp-ad--top-above-nav.
  * @param {HTMLElement} adSlotNode
  */
 const renderAdvertLabel = (adSlotNode: HTMLElement): Promise<Promise<void>> => {
-	let renderDynamic = true;
-	const shouldRender = shouldRenderLabel(adSlotNode);
-
 	return fastdom.measure(() => {
-		if (adSlotNode.id === 'dfp-ad--top-above-nav') {
-			const labelToggle = document.querySelector<HTMLElement>(
-				'.ad-slot__label.ad-slot__label--toggle',
-			);
-			if (labelToggle) {
-				// found a toggled label so don't render dynamically
-				renderDynamic = false;
-				if (shouldRender) {
-					void fastdom.mutate(() => {
-						labelToggle.classList.remove('hidden');
-						labelToggle.classList.add('visible');
-					});
-				} else {
-					// some ads should not have a label
-					// for example fabric ads can have an embedded label
-					// so don't display and remove from layout
-					return fastdom.mutate(() => {
-						labelToggle.style.display = 'none';
-					});
-				}
-			}
-		}
-
-		if (renderDynamic && shouldRender) {
+		if (shouldRenderLabel(adSlotNode)) {
+			const adLabelContent = `Advertisement${createAdTestLabel()}`;
 			return fastdom.mutate(() => {
-				adSlotNode.prepend(createAdLabel());
+				adSlotNode.setAttribute('data-label-show', 'true');
+				adSlotNode.setAttribute('ad-label-text', adLabelContent);
+				adSlotNode.insertBefore(
+					createAdCloseDiv(),
+					adSlotNode.firstChild,
+				);
+				if (shouldRenderAdTestLabel()) {
+					adSlotNode.insertBefore(
+						createAdTestCookieRemovalLink(),
+						adSlotNode.firstChild,
+					);
+				}
```

Comment:
I feel there _might_ be some simplification to be had here.

We are calling `createAdTestLabel` which internally calls `shouldRenderAdTestLabel` which we call again on line 100.

Could we call that early and set a const / pass a flag to the methods that require it?

Also the adtest param could be fetched and passed to methods that require them rather than them each calling `getCookie` - although they are memoised, it could mean we can make the helper methods pure and potentially the flow a bit simpler.

Reply:
That's a good shout, I've tweaked the code so that now, shouldRenderAdTestLabel is only called once and the result is passed into the other function. I've done the same with getCookie - hopefully that's a bit neater!